cpp/oneapi/dal/table/backend/csr_kernels.cpp

```diff
@@ -0,0 +1,212 @@
+        return true;
+    }
+    else {
+        static_assert("has_array_data_kind(): undefined policy type");
+    }
+
+    return false;
+}
+
+template <typename DataSrc, typename DataDest>
+static void refer_origin_data(const array<DataSrc>& src,
+                              std::int64_t src_start_index,
+                              std::int64_t dst_count,
+                              array<DataDest>& dst) {
+    ONEDAL_ASSERT(src_start_index >= 0);
+    ONEDAL_ASSERT(src.get_count() > src_start_index);
+    ONEDAL_ASSERT((src.get_count() - src_start_index) * sizeof(DataSrc) >=
+                  dst_count * sizeof(DataDest));
+
+    if (src.has_mutable_data()) {
+        // TODO: in future, when table knows about mutability of its data this branch shall be
+        // available only for builders, not for tables.
+        auto start_pointer = reinterpret_cast<DataDest*>(src.get_mutable_data() + src_start_index);
+        dst.reset(src, start_pointer, dst_count);
+    }
+    else {
+        auto start_pointer = reinterpret_cast<const DataDest*>(src.get_data() + src_start_index);
+        dst.reset(src, start_pointer, dst_count);
+    }
+}
+
+static void check_origin_data(const array<byte_t>& origin_data,
+                              std::int64_t element_count,
+                              std::int64_t origin_dtype_size,
+                              std::int64_t block_dtype_size) {
+    detail::check_mul_overflow(element_count, std::max(origin_dtype_size, block_dtype_size));
+    ONEDAL_ASSERT(origin_data.get_count() >= element_count * origin_dtype_size);
+}
+
+template <typename Policy, typename BlockData>
+void pull_sparse_block_impl(const Policy& policy,
+                            const csr_info& origin_info,
+                            const block_info& block_info,
+                            const array<byte_t>& origin_data,
+                            const array<std::int64_t>& origin_column_indices,
+                            const array<std::int64_t>& origin_row_indices,
+                            detail::sparse_block<BlockData>& block,
+                            alloc_kind kind) {
+    constexpr std::int64_t block_dtype_size = sizeof(BlockData);
+    const auto origin_dtype_size = detail::get_data_type_size(origin_info.dtype_);
+
+    // overflows checked here
+    check_origin_data(origin_data, origin_info.element_count_, origin_dtype_size, block_dtype_size);
+
+    const auto block_dtype = detail::make_data_type<BlockData>();
+
+    const std::int64_t origin_offset =
+        origin_row_indices[block_info.row_offset_] - origin_row_indices[0];
+
+    const std::int64_t block_size =
+        origin_row_indices[block_info.row_offset_ + block_info.row_count_] -
+        origin_row_indices[block_info.row_offset_];
+
+    const bool same_type(block_dtype == origin_info.dtype_);
+    //const bool same_indexing(block_info.indexing_ == origin_info.indexing_);
+
+    // data
```

Comment:
If you are separating logic for data, column indices and row indices, it is better to extract particular methods and call them here.

Reply:
In my opinion it won't improve the readability of the code

Reply:
Can you please clarify why? This function will be much shorter and it will be easier to understand its flow:

```cpp
template <typename Policy, typename BlockData>
void pull_sparse_block_impl( /* params */ ) {
/*checks*/
pull_data(/*...*/);
pull_row_indices(/*..*/);
pull_column_indices(/*..*/);
}
```
It would be even easier to understand/modify the logic of its logical parts if they are organized like separate functions.

Reply:
Ok, it was modified